---
title: "MIDORI2: A collection of quality controlled, preformatted, and regularly updated reference databases for taxonomic assignment of eukaryotic mitochondrial sequences"
authors:
- Matthieu Leray
- Nancy Knowlton 
- Ryuji J Machida
date: "2022-04-11T00:00:00Z"
doi: "10.1002/edn3.303"

# Schedule page publish date (NOT publication's date).
publishDate: "2022-04-11T00:00:00Z"

# Publication type.
# Legend: 0 = Uncategorized; 1 = Conference paper; 2 = Journal article;
# 3 = Preprint / Working Paper; 4 = Report; 5 = Book; 6 = Book section;
# 7 = Thesis; 8 = Patent
publication_types: ["2"]

# Publication name and optional abbreviated publication name.
publication: "[Environmental DNA (2022) 4; 894–-907](https://doi.org/10.1002/edn3.303)"
publication_short: ""

abstract: Analysis of environmental DNA is increasingly used to characterize ecological communities, but the effectiveness of this approach depends on the accuracy of taxonomic reference databases. The MIDORI databases, first released in 2017, were built to improve accuracy for mitochondrial metazoan (animal) sequences. MIDORI has now been significantly improved and renamed MIDORI2 (available at http://www.reference-midori.info). Like MIDORI, MIDORI2 is built from GenBank and contains curated sequences of thirteen protein-coding and two ribosomal RNA mitochondrial genes. Coverage has been substantially expanded to cover all eukaryotes, including fungi, green algae and land plants, other multicellular algal groups, and diverse protist lineages. MIDORI2 also now includes not only species with full binomials, but also taxa referred to by genus with species left unspecified (“sp.”). Another new feature is the updating of the databases approximately every two months with version numbers corresponding to each new GenBank release. Additional potentially erroneously annotated sequences have also been removed. Finally, the ability to export data files to BLAST+ has been added to the original ability to export preformatted data to five taxonomic assignment programs, and databases of amino acid sequences are also made available for protein-coding genes. As a technical validation, we conducted a preliminary comparison of the performance of MIDORI2 with five taxonomic assignment programs. Results suggest that BLAST+ top hits performed better for assigning CO1 sequences than alignment-free methods based on compositional features. Comparing MIDORI2 with two other commonly used curated databases of mitochondrial sequences, CO-ARBitrator and BOLD, we show that MIDORI2 includes sequences from a broader range of metazoan and non-metazoan taxa. Overall, in many contexts, MIDORI2 offers clear advantages--a higher diversity of taxa than other databases, a variety of user-friendly features, and regular updates. MIDORI2 is particularly well-suited for environmental DNA studies that target mitochondrial genes with broad primers.

# Summary. An optional shortened abstract.
summary:

tags:
- eukaryote
- fungi
- GenBank
- metabarcoding
- metazoan
- mitochondrial
- genes
- plants
- protist

#featured: false

links:
# - name: ""
#   url: ""
- icon: open-access
  icon_pack: ai
  name: Open Access
  url: https://onlinelibrary.wiley.com/doi/full/10.1002/edn3.303
url_pdf: https://onlinelibrary.wiley.com/doi/epdf/10.1002/edn3.303
url_code: ''
url_project:
url_dataset:
url_poster: ''
url_slides: ''
url_source:
url_video: ''

# Featured image
# To use, add an image named `featured.jpg/png` to your page's folder.
image:
  caption: 'publication Figure 2, created by Natalie Renier (http://nrenier.com/), Woods Hole Oceanographic Institution.'
  focal_point: ""
  preview_only: false

# Associated Projects (optional).
#   Associate this publication with one or more of your projects.
#   Simply enter your project's folder or file name without extension.
#   E.g. `internal-project` references `content/project/internal-project/index.md`.
#   Otherwise, set `projects: []`.
projects: []

# Slides (optional).
#   Associate this publication with Markdown slides.
#   Simply enter your slide deck's filename without extension.
#   E.g. `slides: "example"` references `content/slides/example/index.md`.
#   Otherwise, set `slides: ""`.
slides: example
design:
  css_class: 'pub_class'
---

<script type='text/javascript' src='https://d1bxh8uas1mnw7.cloudfront.net/assets/embed.js'></script>

<span data-badge-type="medium-donut" data-doi="10.1002/edn3.303" data-condensed="true" data-hide-no-mentions="true" class="altmetric-embed"></span> <span class="__dimensions_badge_embed__" data-doi="10.1002/edn3.303" data-hide-zero-citations="false" data-legend="hover-right"></span><script async src="https://badge.dimensions.ai/badge.js" charset="utf-8"></script>

<br/>
